jetty-server/src/test/java/org/eclipse/jetty/server/DetectorConnectionFactoryTest.java

```diff
@@ -0,0 +1,584 @@
+        assertThat(response, Matchers.containsString("HTTP/1.1 400"));
+    }
+
+    @Test
+    void testDetectorToNextDetector() throws Exception
+    {
+        String keystore = MavenTestingUtils.getTestResourceFile("keystore").getAbsolutePath();
+        SslContextFactory sslContextFactory = new SslContextFactory.Server();
+        sslContextFactory.setKeyStorePath(keystore);
+        sslContextFactory.setKeyStorePassword("storepwd");
+        sslContextFactory.setKeyManagerPassword("keypwd");
+
+        ProxyConnectionFactory proxy = new ProxyConnectionFactory(HttpVersion.HTTP_1_1.asString());
+        DetectorConnectionFactory proxyDetector = new DetectorConnectionFactory(proxy);
+        SslConnectionFactory ssl = new SslConnectionFactory(sslContextFactory, proxyDetector.getProtocol());
+        DetectorConnectionFactory sslDetector = new DetectorConnectionFactory(ssl);
+        HttpConnectionFactory http = new HttpConnectionFactory();
+
+        start(sslDetector, proxyDetector, http);
+
+        String request = "PROXY TCP 1.2.3.4 5.6.7.8 111 222\r\n" +
+            "GET /path HTTP/1.1\n" +
+            "Host: server:80\n" +
+            "Connection: close\n" +
+            "\n";
+        String response = getResponseOverSsl(request);
+
+        // SSL matched, so the upgrade was made to proxy which itself upgraded to HTTP
+        assertThat(response, Matchers.containsString("HTTP/1.1 200"));
+        assertThat(response, Matchers.containsString("pathInfo=/path"));
+        assertThat(response, Matchers.containsString("local=5.6.7.8:222"));
+        assertThat(response, Matchers.containsString("remote=1.2.3.4:111"));
+    }
+
+    @Test
+    void testDetectorWithDetectionUnsuccessfulLambda() throws Exception
+    {
+        AtomicBoolean detectionSuccessful = new AtomicBoolean(true);
+        ProxyConnectionFactory proxy = new ProxyConnectionFactory(HttpVersion.HTTP_1_1.asString());
+        DetectorConnectionFactory detector = new DetectorConnectionFactory((connector, endPoint, buffer) ->
+        {
+            if (!detectionSuccessful.compareAndSet(true, false))
+                throw new AssertionError("DetectionUnsuccessful callback should only have been called once");
+
+            // omitting this will leak the buffer
+            connector.getByteBufferPool().release(buffer);
+
+            Callback.Completable completable = new Callback.Completable();
+            endPoint.write(completable, ByteBuffer.wrap("No upgrade for you".getBytes(StandardCharsets.US_ASCII)));
+            completable.whenComplete((r, x) -> endPoint.close());
+        }, proxy);
+        HttpConnectionFactory http = new HttpConnectionFactory();
+
+        start(detector, http);
+
+        String request = "GET /path HTTP/1.1\n" +
+            "Host: server:80\n" +
+            "Connection: close\n" +
+            "\n";
+        String response = getResponse(request);
+
+        assertEquals("No upgrade for you\n", response);
```

Comment:
Why there is an extra `\n` at the end?

Reply:
It's an artifact of the `getResponse` method. I agree that it's confusing so I'll clean this up to make both the generated and the expected strings identical.